components/concurrency_manager/src/tracked_arc.rs

```diff
@@ -0,0 +1,1125 @@
+        }
+    }
+
+    /// Check if this instance should be tracked based on sampling rate
+    pub fn should_track(&self) -> bool {
+        let rate = self.sampling_rate.load(Ordering::Relaxed);
+        if rate == 0 {
+            return false; // No tracking when rate is 0
+        }
+        if rate == 1 {
+            return true; // Always track when rate is 1
+        }
+
+        let count = self.sampling_counter.fetch_add(1, Ordering::Relaxed);
+        count % rate == 0
+    }
+
+    /// Record an access event for a KeyHandle
+    /// SAFETY: Avoids deadlock by cloning the Arc before acquiring Mutex
+    pub fn record_access(&self, key_handle_id: u64, access_info: AccessInfo) {
+        // SAFE: Get Arc<TrackedInfo> and release DashMap lock immediately
+        let tracked_info = match self.registry.get(&key_handle_id) {
+            Some(entry) => entry.value().clone(), // Clone Arc, release DashMap lock
+            None => {
+                // the record may have been cleared
+                TRACKED_ARC_RECORD_NOT_FOUND.inc();
+                return;
+            }
+        };
+
+        // Now safely acquire Mutex without holding DashMap lock
+        let mut history = match tracked_info.access_history.lock() {
+            Ok(history) => history,
+            Err(_) => {
+                error!("lock is poisoned");
+                return;
+            }
+        };
+
+        history.push_back(access_info);
+
+        let max_size = self.max_history_size.load(Ordering::Relaxed);
+
+        if history.len() > max_size {
+            history.pop_front();
+        }
+    }
+
+    /// Remove a TrackedInfo from the registry
+    pub fn unregister(&self, id: u64) {
+        self.registry.remove(&id);
+    }
+
+    /// Retrieve a cloned Arc<TrackedInfo> by ID (deadlock-safe)
+    /// SAFETY: Returns cloned Arc, immediately releases DashMap lock
+    pub fn get_info(&self, id: u64) -> Option<Arc<TrackedInfo>> {
+        self.registry.get(&id).map(|entry| entry.value().clone())
+    }
+
+    /// Get summary information about all tracked objects
+    pub fn get_summary(&self) -> Vec<(u64, String, usize)> {
+        self.registry
+            .iter()
+            .map(|entry| entry.value().clone())
+            .map(|info| {
+                let access_count = info.access_history.lock().map(|h| h.len()).unwrap_or(0);
+                (info.key_handle_id, format!("{:?}", info.key), access_count)
+            })
+            .collect()
+    }
+
+    /// Get detailed information about a specific tracked object including
+    /// access history
+    pub fn get_detailed_info(&self, key_handle_id: u64) -> Option<String> {
+        // do not hold references into the dashmap
+        let info: Arc<TrackedInfo> = self.get_info(key_handle_id)?;
+
+        let mut result = format!(
+            "KeyHandle ID: {}\n\
+             Key: {:?}\n",
+            info.key_handle_id, info.key
```

Comment:
Maybe better use hex to format the key

Reply:
The Debug implementation for `Key` already uses the hex format.